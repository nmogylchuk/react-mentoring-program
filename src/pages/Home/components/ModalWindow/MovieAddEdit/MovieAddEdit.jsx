import React from 'react';
import {useFormik} from 'formik';
import * as Yup from 'yup';
import {useDispatch} from 'react-redux';
import ModalWindow from '../ModalWindow';
import InputField from '../../../../shared/InputField/InputField';
import InputSelect from '../../../../shared/InputSelect/InputSelect';
import './MovieAddEdit.scss';
import {addMovie, updateMovie} from '../../../../../store/actions/MoviesActions';
import {genres} from '../../../../../utils/constants';

const MovieAddEdit = ({onClose, movie, isEdit}) => {

    const dispatch = useDispatch();

    const initObject = isEdit ? {
        id: movie.id,
        title: movie.title,
        release_date: movie.date,
        genres: movie.genres,
        overview: movie.overview,
        runtime: movie.runtime == null ? 0 : movie.runtime,
        poster_path: movie.poster_path,
    } : {
        title: '',
        release_date: new Date().toISOString().substr(0, 10),
        genres: genres,
        overview: '',
        runtime: 0,
        poster_path: 'http://test_path.com/'
    };
    const requiredErrorMessage = 'This field is required';
    const formik = useFormik({
        initialValues: initObject,

        validationSchema: Yup.object({
            title: Yup.string().required(requiredErrorMessage),
            release_date: Yup.string().required(requiredErrorMessage),
            genres: Yup.array().required(requiredErrorMessage),
            overview: Yup.string().required(requiredErrorMessage),
            runtime: Yup.number().min(0).required(requiredErrorMessage),
            poster_path: Yup.string().url().required(requiredErrorMessage),
        }),

        onSubmit: values => {
            if (isEdit) {
                dispatch(updateMovie({...values}));
            } else {
                dispatch(addMovie({...values}));
            }
            onClose();
        },

    });

    return (
        <ModalWindow closeWindow={onClose}>
            <form className="movie-edit__form" onSubmit={formik.handleSubmit}>
                <InputField label="Title" value={formik.values.title} type="text" name="title"
                            placeholder="Title here"
                            handleMovieChange={formik.handleChange}/>
                {formik.errors.title ? <div>{formik.errors.title}</div> : null}

                <InputField label="Release date" value={formik.values.release_date} type="date" name="release_date"
                            placeholder="Release date here"
                            handleMovieChange={formik.handleChange}/>
                {formik.errors.release_date ? <div>{formik.errors.release_date}</div> : null}

                <InputSelect label="Genres" name="genres" genres={genres}
                             value={formik.values.genres}
                             handleMovieChange={formik.handleChange}/>
                {formik.errors.genres ? <div>{formik.errors.genres}</div> : null}

                <InputField label="Overview" value={formik.values.overview} type="text" name="overview"
                            placeholder="Overview here"
                            handleMovieChange={formik.handleChange}/>
                {formik.errors.overview ? <div>{formik.errors.overview}</div> : null}

                <InputField label="Runtime" value={formik.values.runtime} type="number" name="runtime"
                            placeholder="Runtime here"
                            handleMovieChange={formik.handleChange}/>
                {formik.errors.runtime ? <div>{formik.errors.runtime}</div> : null}

                <div className="movie-edit__button-wrapper">
                    <button className="movie-edit__button movie-edit__button--reset" type="reset">Reset
                    </button>
                    <button className="movie-edit__button movie-edit__button--submit"
                            type="submit">{isEdit ? 'Save' : 'Submit'}
                    </button>
                </div>
            </form>
        </ModalWindow>
    );
}

export default MovieAddEdit;
