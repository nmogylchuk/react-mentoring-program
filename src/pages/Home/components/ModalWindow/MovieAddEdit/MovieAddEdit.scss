@import 'src/assets/styles/variables';
@import 'src/assets/styles/mixins';

.movie-edit {
  &__title {
    font-size: 2rem;
    text-transform: uppercase;
  }

  &__button {
    @include button();

    &-wrapper {
      display: flex;
      justify-content: flex-end;
      margin-top: 20px;
    }

    &--reset {
      background: $dark-grey;
      color: $red;
      border: 1px solid $red;
      margin-right: 10px;

      &:hover {
        background: $red;
        color: $white;
      }
    }

    &--submit {
      background: $red;
      color: $white;

      &:hover {
        background: $dark-grey;
        color: $red;
        border: 1px solid $red;
      }
    }
  }
}
