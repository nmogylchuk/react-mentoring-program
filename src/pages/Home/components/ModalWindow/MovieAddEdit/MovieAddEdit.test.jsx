import React from 'react';
import MovieAddEdit from "./MovieAddEdit";
import rerender from "react-test-renderer";
import { jest } from "@jest/globals";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

describe('when Movie Add Edit', () => {

    const movie = {
        "id": 447365,
        "title": "Guardians of the Galaxy Vol. 3",
        "tagline": "",
        "vote_average": 0,
        "vote_count": 9,
        "release_date": "2020-05-01",
        "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
        "overview": "The third film based on Marvel's Guardians of the Galaxy.",
        "budget": 0,
        "revenue": 0,
        "genres": [
            "Action",
            "Adventure",
            "Science Fiction"
        ],
        "runtime": null
    }

    test('then snapshot created', () => {
        const initialStore = {};
        const store = configureStore([])(initialStore);
        store.dispatch = jest.fn();
        const movieAddEditComponent = rerender.create(<Provider store={store}>
            <MovieAddEdit onClose={jest.fn()} movie={movie} isEdit={true}/>
        </Provider>);

        const tree = movieAddEditComponent.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('then snapshot update', () => {
        const initialStore = {};
        const store = configureStore([])(initialStore);
        store.dispatch = jest.fn();
        const movieAddEditComponent = rerender.create(<Provider store={store}>
            <MovieAddEdit onClose={jest.fn()} movie={movie} isEdit={false}/>
        </Provider>);

        const tree = movieAddEditComponent.toJSON();
        expect(tree).toMatchSnapshot();
    });
});