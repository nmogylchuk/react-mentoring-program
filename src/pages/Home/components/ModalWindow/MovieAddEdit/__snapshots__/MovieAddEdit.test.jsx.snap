// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when Movie Add Edit then snapshot created 1`] = `null`;

exports[`when Movie Add Edit then snapshot update 1`] = `null`;
